csi/controller_server.go

```diff
@@ -370,64 +407,52 @@ func (cs *ControllerServer) ControllerUnpublishVolume(ctx context.Context, req *
 	// VOLUME_NOT_FOUND is no longer the ControllerUnpublishVolume error
 	// See https://github.com/container-storage-interface/spec/issues/382 for details
 	if existVol == nil {
-		msg := fmt.Sprintf("ControllerUnpublishVolume: the volume %s not exists", req.GetVolumeId())
-		logrus.Warn(msg)
+		logrus.Warnf("ControllerUnpublishVolume: the volume %s does not exists", req.GetVolumeId())
+		return &csi.ControllerUnpublishVolumeResponse{}, nil
+	}
+
+	if existVol.State == string(types.VolumeStateDetached) {
+		logrus.Infof("don't need to detach Volume %s since we are already detached from node %s",
+			req.GetVolumeId(), req.GetNodeId())
 		return &csi.ControllerUnpublishVolumeResponse{}, nil
 	}
 
 	if existVol.State == string(types.VolumeStateDetaching) {
 		return nil, status.Errorf(codes.Aborted, "The volume %s is detaching", req.GetVolumeId())
 	}
 
-	needToDetach := false
-	if existVol.State == string(types.VolumeStateAttached) || existVol.State == string(types.VolumeStateAttaching) {
-
-		// try to wait till we are attached, but if we fail to attach we need to process the detach
-		// irregardless of the node we are on otherwise we might end up, allowing a bugged volume to remain in the attaching state forever
-		// NOTE: this is a tradeoff and it would be better if we could verify that we are actually stuck attaching to the requested node
-		if existVol.State == string(types.VolumeStateAttaching) {
-			if cs.waitForVolumeState(req.GetVolumeId(), string(types.VolumeStateAttached), isVolumeAttached, false, true) {
-				existVol, err = cs.apiClient.Volume.ById(req.GetVolumeId())
-				if err != nil {
-					return nil, status.Error(codes.Internal, err.Error())
-				}
-			} else {
-				logrus.Warnf("Volume %s stuck in attaching state, processing detach request for node %s "+
-					"even though we don't know which node we are attaching on", req.GetVolumeId(), req.GetNodeId())
-			}
+	// try to wait till we are attached, but if we fail to attach we need to process the detach
+	// irregardless of the node we are on otherwise we might end up, allowing a bugged volume to remain in the attaching state forever
+	// NOTE: this is a tradeoff and it would be better if we could verify that we are actually stuck attaching to the requested node
```

Comment:
~~Is this refactoring related to the live migration feature? If NO, you can launch the separate PR or commit then.~~

I saw you removed this part in the last commit. Can you directly remove this part in this commit instead?

Reply:
This was a code move,
I added this code a while back to protect against detach requests which would nuke a valid attachment, since we had no way of knowing which node we were attached to.

Now that we no longer need this, I remove it in a follow on commit.
Generally I like using smaller commits which explain why we are doing something, If I add it to the previous commit, that information is lost.

How about I move the removal commit right next to the attach/detach commit?

Reply:
I squashed all the separate changes into one commit :)